pkg/ccl/cloudccl/amazon/amazon_test: TestAWSKMSExternalConnection failed

pkg/ccl/cloudccl/amazon/amazon_test.TestAWSKMSExternalConnection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/20445838?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/20445838?buildTab=artifacts#/) on master @ [b26650a4c5ddb781eec00f4e06f9f48657427f83](https://github.com/cockroachdb/cockroach/commits/b26650a4c5ddb781eec00f4e06f9f48657427f83):


```
=== RUN   TestAWSKMSExternalConnection
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/90c1c2a76fdb8bf273c5e978243dfe5a/logTestAWSKMSExternalConnection1761983997
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    s3_connection_test.go:224: error executing query="RESTORE DATABASE foo FROM LATEST IN 'external://backup' WITH new_db_name = bar, kms='external://auth-implicit-kms'" args=[]: pq: getting backup index metadata: reading index file 82fe66d42ff673_0_20250909-001118.28_metadata.pb bytes: EOF
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/90c1c2a76fdb8bf273c5e978243dfe5a/logTestAWSKMSExternalConnection1761983997
--- FAIL: TestAWSKMSExternalConnection (3.00s)
=== RUN   TestAWSKMSExternalConnection/auth-implicit
    testing.go:1577: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
--- FAIL: TestAWSKMSExternalConnection/auth-implicit (1.51s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAWSKMSExternalConnection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54274

pkg/ccl/cloudccl/amazon/amazon_test.TestAWSKMSExternalConnection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/20466895?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/20466895?buildTab=artifacts#/) on master @ [8b39430fe2e0760df13d12a2c95e91600aff4efe](https://github.com/cockroachdb/cockroach/commits/8b39430fe2e0760df13d12a2c95e91600aff4efe):


```
=== RUN   TestAWSKMSExternalConnection
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/90c1c2a76fdb8bf273c5e978243dfe5a/logTestAWSKMSExternalConnection3294777880
    test_log_scope.go:82: use -show-logs to present logs inline
    s3_connection_test.go:224: error executing query="RESTORE DATABASE foo FROM LATEST IN 'external://backup' WITH new_db_name = bar, kms='external://auth-implicit-kms'" args=[]: pq: getting backup index metadata: reading index file 82fe66ba741035_0_20250914-000702.85_metadata.pb bytes: EOF
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/90c1c2a76fdb8bf273c5e978243dfe5a/logTestAWSKMSExternalConnection3294777880
--- FAIL: TestAWSKMSExternalConnection (1.88s)
=== RUN   TestAWSKMSExternalConnection/auth-implicit
    testing.go:1577: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
--- FAIL: TestAWSKMSExternalConnection/auth-implicit (1.27s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAWSKMSExternalConnection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

pkg/ccl/cloudccl/amazon/amazon_test.TestAWSKMSExternalConnection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/20477688?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_CloudUnitTests/20477688?buildTab=artifacts#/) on master @ [5154683affd4b9afa0d8211cc503191679d4cc62](https://github.com/cockroachdb/cockroach/commits/5154683affd4b9afa0d8211cc503191679d4cc62):


```
=== RUN   TestAWSKMSExternalConnection
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/90c1c2a76fdb8bf273c5e978243dfe5a/logTestAWSKMSExternalConnection1842699093
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:154: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    s3_connection_test.go:224: error executing query="RESTORE DATABASE foo FROM LATEST IN 'external://backup' WITH new_db_name = bar, kms='external://auth-implicit-kms'" args=[]: pq: getting backup index metadata: reading index file 82fe66aafffa35_0_20250917-000808.90_metadata.pb bytes: EOF
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/90c1c2a76fdb8bf273c5e978243dfe5a/logTestAWSKMSExternalConnection1842699093
--- FAIL: TestAWSKMSExternalConnection (2.42s)
=== RUN   TestAWSKMSExternalConnection/auth-implicit
    testing.go:1577: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
--- FAIL: TestAWSKMSExternalConnection/auth-implicit (1.29s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAWSKMSExternalConnection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/backup/backupdest/backup_index.go
+++ b/pkg/backup/backupdest/backup_index.go
@@ -25,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/ioctx"
 	"github.com/cockroachdb/cockroach/pkg/util/metamorphic"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
@@ -239,7 +240,7 @@ func GetBackupTreeIndexMetadata(
 			if err != nil {
 				return err
 			}
-			reader, size, err := store.ReadFile(
+			reader, _, err := store.ReadFile(
 				ctx, path.Join(indexDir, basename), cloud.ReadOptions{},
 			)
 			if err != nil {
@@ -247,9 +248,9 @@ func GetBackupTreeIndexMetadata(
 			}
 			defer reader.Close(ctx)
 
-			bytes := make([]byte, size)
-			if _, err := reader.Read(ctx, bytes); err != nil {
-				return errors.Wrapf(err, "reading index file %s bytes", basename)
+			bytes, err := ioctx.ReadAll(ctx, reader)
+			if err != nil {
+				return errors.Wrapf(err, "reading index file %s", basename)
 			}
 
 			index := backuppb.BackupIndexMetadata{}

